agent/src/main/scala/za/co/absa/atum/agent/AtumAgent.scala

```diff
@@ -47,25 +47,41 @@ class AtumAgent private() {
    *  @return
    */
   def getOrCreateAtumContext(atumPartitions: AtumPartitions): AtumContext = {
-    contexts.getOrElse(atumPartitions, new AtumContext(atumPartitions, this))
+    getContextOrElse(
+      atumPartitions,
+      AtumContext.fromDTO(
+        dispatcher.getOrCreateAtumContext(AtumPartitions.toPartitioning(atumPartitions), None),
+        this
+      )
+    )
   }
 
   def getOrCreateAtumSubContext(subPartitions: AtumPartitions)(implicit parentAtumContext: AtumContext): AtumContext = {
     val newPartitions: AtumPartitions = parentAtumContext.atumPartitions ++ subPartitions
-    getContextOrElse(newPartitions, parentAtumContext.copy(atumPartitions = newPartitions, agent = this))
+    getContextOrElse(
+      newPartitions,
+      AtumContext.fromDTO(
```

Comment:
potentially same comment as above, it can be extracted into val and not nested like this for code readability

Reply:
Refactored.